File: python/50-Pow/main.py
import unittest

class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n == 0 or x == 1:
            return 1

        if n == 1:
            return x

        if x == 0:
            return 0

        result = 1
        scaling = abs(n)
        is_negative = (n < 0)
        
        for i in range(abs(n)):
            if is_negative:
                result /= x
            else:
                result *= x

        return result

class Tests(unittest.TestCase):
    def test_example1(self):
        solution = Solution()
        self.assertEqual(solution.myPow(2, 10), 1024)

    def test_example2(self):
        solution = Solution()
        self.assertAlmostEqual(solution.myPow(2.1, 3), 9.2610)

    def test_example3(self):
        solution = Solution()
        self.assertAlmostEqual(solution.myPow(2.0, -2), 0.2500)

    def test_example4(self):
        solution = Solution()
        self.assertAlmostEqual(solution.myPow(-2, 2), 4)

if __name__ == '__main__':
    unittest.main()

File: go/50-Pow/main.go
package main

import "fmt"

func myPow(x float64, n int) float64 {
	switch {
	case x == 1 || n == 0:
		return 1

	case n == 1:
		return x

	case n == -1:
		return 1 / x

	case x == 0:
		return 0
	}

	half := myPow(x, n/2)
	result := half * half

	if n&1 == 1 {
		sign := 1

		if n < 0 {
			sign = -1
		}

		return result * myPow(x, sign)
	} else {
		return result
	}
}

func main() {
	fmt.Println(myPow(2.0, 10))
	fmt.Println(myPow(2.1, 3))
	fmt.Println(myPow(2.0, -2))
	fmt.Println(myPow(2.0, 10))
}
